### PR TITLE
fix(dec-20-audit): [L06] Inconsistent hasPrice determination

### PR DESCRIPTION
**Pull request by daywiss**
Signed-off-by: David Adams <david@umaproject.org>

**Motivation**

[L06] Inconsistent hasPrice determination
The hasPrice function of the OptimisticOracle contract returns true if the request is either Settled or Resolved. Although the internal state of the contract does not record the price in the Resolved state, this makes sense because retrieving the price will settle the request. This means that from a user’s point of view, a Resolved price request can be considered to have a price. However, the same reasoning suggests that an Expired price request can also be considered to have a price.

Consider adding another condition to the hasPrice function so it returns true for Expired price requests.

**Summary**

This adds an additional state check to resolve hasPrice to true if price proposal is also expired. Adds in several tests to make sure hasPrice works as intended in various scenarios, including the scenario indicated above where price proposal is expired but not settled.

### CODE REVIEW

File: packages/core/test/oracle/OptimisticOracle.js [hunk 1]
```diff
@@ -154,6 +154,72 @@ contract("OptimisticOracle", function(accounts) {
...
+
+      await collateral.approve(optimisticOracle.address, totalDefaultBond, { from: disputer });
+      await optimisticOracle.disputePrice(optimisticRequester.address, identifier, requestTime, "0x", {
+        from: disputer
+      });
+
+      const result = await optimisticOracle.hasPrice(optimisticRequester.address, identifier, requestTime, "0x");
+      assert.equal(result, false);
+    });
+    it("Should return true when price is proposed and past liveness but not settled", async function() {
```
Comment (nicholaspai) [inline packages/core/test/oracle/OptimisticOracle.js:197]: Isn't this a duplicate test of the one beginning on line 174?

File: packages/core/test/oracle/OptimisticOracle.js [hunk 2]
```diff
@@ -154,6 +154,72 @@ contract("OptimisticOracle", function(accounts) {
...
+      await collateral.approve(optimisticOracle.address, totalDefaultBond, { from: proposer });
+      await optimisticOracle.proposePrice(optimisticRequester.address, identifier, requestTime, "0x", correctPrice, {
+        from: proposer
+      });
+      await timer.setCurrentTime((await timer.getCurrentTime()).add(await optimisticOracle.defaultLiveness()));
+      const result = await optimisticOracle.hasPrice(optimisticRequester.address, identifier, requestTime, "0x");
+      assert.equal(result, true);
+    });
+    it("Should return true when price is proposed, disputed and resolved by dvm", async function() {
```
Comment (nicholaspai) [inline packages/core/test/oracle/OptimisticOracle.js:206]: Do you think you should also test the path where the price is proposed, past liveness, and settled?

**Review by nicholaspai — COMMENTED** [review]
I'm surprised we didn't have tests for `hasPrice` before, thanks for adding! I left some nits on your tests

**Review by mrice32 — APPROVED** [review]
LGTM once @nicholaspai's comments are resolved!

**Review by chrismaree — APPROVED** [review]

File: packages/core/test/oracle/OptimisticOracle.js (hunk 1 above)
Comment (daywiss) [reply packages/core/test/oracle/OptimisticOracle.js:197]: could be, i will check it out

Comment (daywiss) [reply packages/core/test/oracle/OptimisticOracle.js:197]: yes you are right, thanks good catch

**Pull request merged**